lib/Driver/Compilation.cpp

```diff
@@ -600,20 +809,39 @@ namespace driver {
                     std::bind(&PerformJobsState::taskSignalled, this,
                               _1, _2, _3, _4, _5, _6));
 
-        // Mark all remaining deferred commands as skipped.
+        // Returning from TaskQueue::execute should mean either an empty
+        // TaskQueue or a failed subprocess.
+        assert(!(Result == 0 && TQ->hasRemainingTasks()));
+
+        // Task-exit callbacks from TaskQueue::execute may have unblocked jobs,
+        // which means there might be PendingExecution jobs to enqueue here. If
+        // there are, we need to continue trying to make progress on the
+        // TaskQueue before we start marking deferred jobs as skipped, below.
+        if (!PendingExecution.empty() && Result == 0) {
+          formBatchJobsAndAddPendingJobsToTaskQueue();
```

Comment:
I suppose if this still bugs you another option is to put it at the *top* of the loop.

Reply:
Before the call to TaskQueue::execute? I .. am pretty sure that wouldn't work.

Reply:
(It doesn't bug me enough to carry on fiddling it further; I think there are about 2 variants that work right and they're similarly confusing. One has two calls, the other needs more explanation, *shrug*)

Reply:
Yeah, it would also require changing the loop's controlling condition. This is fine.